## **Modelo de Preaprobacion**
* **Caso de estudio:** Se quiere implementar un modelo de preaprobacion de creditos a partir de las variables socioeconommicas
Estas variables son:
* Edad
*Ingresos
*Egresos
*Monto (EAD)


La variable de referencia sera la variabale Preaprobacion. Esta indica la pre aprobacion o pre negacion de un credito.


1. Se cargan las librerias de trabajo

In [ ]:
# %load_ext cudf.pandas
import pandas as pd
import numpy as np
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix

2. Cargar datos de trabajo

In [ ]:
nxl='/content/sample_data/6. SolicitantesCrédito(USD) (1).xlsx'
XDB=pd.read_excel(nxl,sheet_name=0) #DataFrame : Tabla de datos - Tablas de Word

#Seleccionamos las variables de trabajo
XDB=XDB[['Edad','Ingresos','Egresos','Monto (EAD)','PreApr']]
filas=np.where(~np.isnan(XDB['PreApr']))[0]
XDB=XDB.iloc[filas,:]
XDB=XDB.dropna()


#Separar variables de entrada - variables de salida
XD=XDB[['Edad','Ingresos','Egresos','Monto (EAD)']]
yd=XDB['PreApr']#Variables de Salida
display(XDB.head())

,Edad,Ingresos,Egresos,Monto (EAD),PreApr
0,38,1356.14400,1685.622667,625.562230,1.0
1,51,286.01600,364.462000,140.031984,1.0
2,37,517.46325,629.208889,284.564492,0.0
3,29,473.27000,551.616889,309.647588,1.0
4,42,750.09175,806.715778,500.663578,1.0


3. **Entrenar el modelo Naive Bayes**

In [ ]:
# Suppress scientific notation
np.set_printoptions(suppress=True)
pd.set_option('display.float_format', lambda x: '%.3f' % x)

In [ ]:
#Crear el Modelo
mnb=GaussianNB()

#Entrenar el Modelo
mnb.fit(XD,yd)

# Datos por categoria
ndat=mnb.class_count_
print('Los datos por categoria son:\n',ndat)
cat=mnb.classes_
print('Las categorias son:\n',cat)

#Determinamos las propiedad estadisticas de las variables
u=mnb.theta_ #La media de variables por categoria
print('La Media de variables es:\n:',u)
sigma=np.sqrt(mnb.var_) #La desviacion estandar de variables por categoria
print('La desviacion estandar de variables es:\n:',sigma)
print('Los limites superiores de las variables son:\n',u+sigma)
print('Los limites inferiores de las variables son:\n',u-sigma)



Los datos por categoria son:
 [2959. 2883.]
Las categorias son:
 [0. 1.]
La Media de variables es:
: [[  40.98411626  431.42680644  401.54995644  289.41445941]
 [  41.10752688 1029.98947251  900.19312144  781.2021157 ]]
La desviacion estandar de variables es:
: [[ 12.21303268 325.79791497 342.77286087 235.95783541]
 [ 12.25106493 472.43342807 525.43337837 469.90038088]]
Los limites superiores de las variables son:
 [[  53.19714893  757.22472141  744.32281731  525.37229482]
 [  53.35859181 1502.42290058 1425.62649981 1251.10249658]]
Los limites inferiores de las variables son:
 [[ 28.77108358 105.62889147  58.77709557  53.456624  ]
 [ 28.85646196 557.55604444 374.75974307 311.30173482]]


4. Se procede con la evaluación de desempeño del modelo( frente al pronostico de la decision)

Falsos: Equivocacion

Verdaderos: Acierto



In [ ]:
ydp=mnb.predict(XD)
cm=confusion_matrix(yd,ydp)
print('La matriz de confusion es:\n',cm)

La matriz de confusion es:
 [[2502  457]
 [ 937 1946]]


5. Metricas de desempeño

In [ ]:
cm=confusion_matrix(yd,ydp)
print('La matriz de confusion es:\n',cm)
VN=cm[0,0]
FP=cm[0,1]
FN=cm[1,0]
VP=cm[1,1]
TD=len(XD) #Total datos

#Exactitud: Muestra el comportamiento general del modelo
Ex=(VP+VN)/TD
print('La exactitud del modelo es:\n',Ex)

#Tasa de error:Porcentaje de equivocacion
ter=(FP+FN)/TD
print('La tasa de error del modelo es:\n',ter)

#Sensibilidad: Como se comporta el modelo frente a los pre aprobados
Se=VP/(VP+FN)
print('La sensibilidad del modelo es:\n',Se)

#Especificidad: Como se comporta el modelo frente a los pre negados
Esp=VN/(VN+FP)
print('La especificidad del modelo es:\n',Esp)

#Precisión: Capacidad del modelo para predecir VP
Pre=VP/(VP+FP)
print('La precision del modelo es:\n',Pre)

#Precision Negativa: Capacidad del modelo para predecir VN
PreN=VN/(VN+FN)
print('La precision negativa del modelo es:\n',PreN)

La matriz de confusion es:
 [[2502  457]
 [ 937 1946]]
La exactitud del modelo es:
 0.7613830879835672
La tasa de error del modelo es:
 0.23861691201643273
La sensibilidad del modelo es:
 0.6749913284772806
La especificidad del modelo es:
 0.8455559310577898
La precision del modelo es:
 0.8098210570120683
La precision negativa del modelo es:
 0.7275370747310265


# **Analisis de Resultados**
* De acuerdo con la base de datos hay un total de 5482 datos, los cuales estan
divididos en 2959 en pre negados y 2883 en pre aprobados.

* Se puede observar igualmente que la categoria de pre aprobados presento los valores mas altos frente a las variable Edad, Ingresos, Egresos, Monto.

* De la matriz de confusion se puede observar que el modelo tiende a equivocarse en mayor medida en el pronostico de los valores pre apobados con un total de 947 falsos negativos, frente  a 457 de Falsos Positivos

* Con las metricas de desempeño basadas en la matriz de confusion, podemos observar que la exactitud del modelo es del 76% lo que muestra el buen comportamiento del modelo frente al pronostico de los datos en general,la tasa de error del modelo es de 24%

* Estas metricas tambien nos muestran que la sensibilidad del modelo a pre aprobar creditos es relativamente baja 67%, mientras la especificidad para negarlos es de 84% por lo tanto se comporta mejor diagnosticando los negativos

* En la precision podemos ver que es mas preciso a pre aprobarlos con una precision del 81%, mientras la precision al momento de pre negarlos es del 73%

6. Evaluar solicitantes de credito

In [ ]:
XI=[35,600,900,300]
ydpi=mnb.predict([XI])
print('La categoria del solicitante es:',ydpi)

La categoria del solicitante es: [0.]


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


7. Creamos con los que no poseen pre aprobacion

In [ ]:
nxl='/content/sample_data/6. SolicitantesCrédito(USD) (1).xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)


XDB=XDB[['Edad','Ingresos','Egresos','Monto (EAD)','PreApr']]
filas_sp=np.where(np.isnan(XDB['PreApr']))[0]
print(filas_sp)
# Seleccionar los datos de los solicitantes sin un valor de PreApr
XDBsp = XDB.iloc[filas_sp,:]
XDBsp = XDBsp.drop(columns=['PreApr'])

# Predecir el estado de PreApr para estos solicitantes
ydsp = mnb.predict(XDBsp)

#Pegamos el pronostico
TAsq=np.column_stack((filas_sp,XDBsp,ydsp))
df=pd.DataFrame(TAsq,columns=['Solicitante','Edad','Ingresos','Egresos','Monto (EAD)','PreApr'])
display(df)

df.to_excel('Resultados_PreApr.xlsx')

[   6    7   11 ... 8367 8372 8377]


,Solicitante,Edad,Ingresos,Egresos,Monto (EAD),PreApr
0,6.000,38.000,383.678,368.599,175.847,0.000
1,7.000,47.000,34.188,38.034,26.956,0.000
2,11.000,46.000,1774.305,2116.724,705.435,1.000
3,13.000,37.000,447.997,309.207,518.676,0.000
4,16.000,22.000,723.136,794.865,419.475,0.000
...,...,...,...,...,...,...
2531,8357.000,24.000,608.292,337.400,776.458,0.000
2532,8358.000,66.000,20.921,26.856,8.495,0.000
2533,8367.000,32.000,283.266,170.266,225.300,0.000
2534,8372.000,35.000,768.288,847.264,479.164,0.000
